### PR TITLE
Add tree field to scene

### PR DESCRIPTION
**Pull request by jrmi**

### CODE REVIEW

File: terra_layer/models.py [hunk 1]
```diff
@@ -12,21 +12,61 @@
...
         max_length=255, upload_to="scene-icons", null=True, default=None
     )
     order = models.IntegerField(default=0, db_index=True)
+    tree = JSONField(default=list)
 
     def get_absolute_url(self):
         return reverse("scene-detail", args=[self.pk])
 
+    def tree2models(self, current_node=None, parent=None, order=0):
+
```
Comment (SebCorbin) [inline terra_layer/models.py]: Document this :)

File: terra_layer/models.py [hunk 2]
```diff
@@ -12,21 +12,61 @@
...
+            current_node = self.tree
+            self.layer_groups.all().delete()  # Clear all groups to generate brand new one
+
+        if not parent:
+            # Create a default unique parent group that is ignored at export
+            parent = LayerGroup.objects.create(view=self, label="Root")
+
+        if isinstance(current_node, list):
+            for idx, child in enumerate(current_node):
+                self.tree2models(current_node=child, parent=parent, order=idx)
```
Comment (SebCorbin) [inline terra_layer/models.py]: Is it possible to add layer at level 0?

File: terra_layer/models.py [hunk 3]
```diff
@@ -47,14 +87,14 @@ class LayerGroup(models.Model):
 
     class Meta:
         unique_together = ["view", "label", "parent"]
-        ordering = ["order"]
+        ordering = ["order", "label"]
```
Comment (SebCorbin) [inline terra_layer/models.py]: Really necessary? Order is set by `tree2models()` already

File: terra_layer/models.py [hunk 4]
```diff
@@ -12,21 +12,61 @@
...
+                label=current_node["title"],
+                exclusive=current_node.get("exclusive", False),
+                order=order,
+            )
+
+            self.tree2models(current_node=current_node["children"], parent=group)
+
+        elif "geolayer" in current_node:
+            # Handle layers
+            layer = Layer.objects.get(pk=current_node["geolayer"])
```
Comment (SebCorbin) [inline terra_layer/models.py]: - What if layer does not exists (race condition) ?
- What if layer is already referenced in another group (duplicate in the tree) ?

File: terra_layer/models.py (hunk 2 above)
Comment (jrmi) [reply terra_layer/models.py]: Now it is !

File: terra_layer/models.py (hunk 3 above)
Comment (jrmi) [reply terra_layer/models.py]: It can't hurt.

File: terra_layer/models.py (hunk 4 above)
Comment (jrmi) [reply terra_layer/models.py]: Should be forbidden in both cases, I add some cheks.

File: terra_layer/models.py (hunk 3 above)
Comment (SebCorbin) [reply terra_layer/models.py]: It really is not necessary, and as the groups will eventually be dropped, no need to add noise in this PR. Also it hurts :)

> Ordering is not a free operation. Each field you add to the ordering incurs a cost to your database. Each foreign key you add will implicitly include all of its default orderings as well.

from https://docs.djangoproject.com/en/2.2/ref/models/options/#django.db.models.Options.ordering

I'm okay with the one from `Layer` though

File: terra_layer/migrations/0047_auto_20191211_1016.py [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("terra_layer", "0046_auto_20191210_1437"),
+    ]
+
+    operations = [
+        migrations.AlterUniqueTogether(name="layergroup", unique_together=set(),),
+    ]
```
Comment (SebCorbin) [inline terra_layer/migrations/0047_auto_20191211_1016.py]: Can you squash all migrations before merging?

File: terra_layer/models.py [hunk 6]
```diff
@@ -107,6 +150,9 @@ def save(self, **kwargs):
         if self.group:
             cache.delete(get_layer_group_cache_key(self.group.view))
 
+    def __str__(self):
+        return f"Layer<{self.id}> - {self.name}"
```
Comment (SebCorbin) [inline terra_layer/models.py]: I'd prefer to use parenthesis here, to keep consistency with Django default behavior which is

```python
    def __repr__(self):
        return '<%s: %s>' % (self.__class__.__name__, self)

    def __str__(self):
        return '%s object (%s)' % (self.__class__.__name__, self.pk)

File: terra_layer/models.py (hunk 3 above)
Comment (jrmi) [reply terra_layer/models.py]: Ok ;-) I remove the one in layer group

File: terra_layer/migrations/0047_auto_20191211_1016.py (hunk 5 above)
Comment (jrmi) [reply terra_layer/migrations/0047_auto_20191211_1016.py]: Yes a will do it later in some time after this message :D

File: terra_layer/models.py (hunk 6 above)
Comment (jrmi) [reply terra_layer/models.py]: You mean the `%` operator ? If we were in django code i'll agree, but here it's our code. I don't see the point to respect an old convention.

Comment (SebCorbin) [reply terra_layer/models.py]: No I mean the `<pk>` instead of `(pk)`

Comment (jrmi) [reply terra_layer/models.py]: Okkkkk ! So I agree !

File: terra_layer/models.py (hunk 1 above)
Comment (jrmi) [reply terra_layer/models.py]: Done

File: terra_layer/models.py (hunk 6 above)
Comment (jrmi) [reply terra_layer/models.py]: done

**Comment by SebCorbin** [comment]
I just saw that we have a `in_tree` field on layer, which is set to True and never changes, now would be a good time to make it work, don't you think?

**Comment by jrmi** [comment]
> I just saw that we have a `in_tree` field on layer, which is set to True and never changes, now would be a good time to make it work, don't you think?

Or to remove this field ?

**Comment by SebCorbin** [comment]
> > I just saw that we have a `in_tree` field on layer, which is set to True and never changes, now would be a good time to make it work, don't you think?
> 
> Or to remove this field ?

I'd say this field might be handy for when we drop groups, as we could identify quickly which layers are "orphan" (and for example sort by this field for a quick layer selection in the tree)

**Comment by jrmi** [comment]
> > > I just saw that we have a `in_tree` field on layer, which is set to True and never changes, now would be a good time to make it work, don't you think?
> > 
> > 
> > Or to remove this field ?
> 
> I'd say this field might be handy for when we drop groups, as we could identify quickly which layers are "orphan" (and for example sort by this field for a quick layer selection in the tree)

Right. I prefer a foreign key between the layer and the view. Today the "foreign" key is already "simulated" and used in admin. Should be sufficient, what do you think ?

**Comment by SebCorbin** [comment]
> > > > I just saw that we have a `in_tree` field on layer, which is set to True and never changes, now would be a good time to make it work, don't you think?
> > > 
> > > 
> > > Or to remove this field ?
> > 
> > 
> > I'd say this field might be handy for when we drop groups, as we could identify quickly which layers are "orphan" (and for example sort by this field for a quick layer selection in the tree)
> 
> Right. I prefer a foreign key between the layer and the view. Today the "foreign" key is already "simulated" and used in admin. Should be sufficient, what do you think ?

I'm okay with that then, so a layer can be attached to one only view, right?

**Comment by jrmi** [comment]
> I'm okay with that then, so a layer can be attached to one only view, right?

Right

**Pull request merged**